crates/subspace-farmer/src/archiving.rs

```diff
@@ -62,96 +68,126 @@ impl Archiving {
-                        object_mapping,
-                    } = archived_segment;
-
-                    let segment_index = root_block.segment_index();
-                    if last_archived_segment_index == Some(segment_index) {
-                        continue;
-                    }
-                    last_archived_segment_index.replace(segment_index);
-
-                    let piece_index_offset = merkle_num_leaves * segment_index;
+        let mut plot_segment = move |archived_segment| {
+            let ArchivedSegment {
+                root_block,
+                pieces,
+                object_mapping,
+            } = archived_segment;
+            let segment_index = root_block.segment_index();
+            let piece_index_offset = merkle_num_leaves * segment_index;
+
+            let pieces_to_plot = PiecesToPlot {
+                piece_indexes: (piece_index_offset..).take(pieces.count()).collect(),
+                pieces,
+            };
+            if !on_pieces_to_plot(pieces_to_plot) {
+                // No need to continue
+                return ControlFlow::Break(());
+            }
 
-                    let pieces_to_plot = PiecesToPlot {
-                        piece_indexes: (piece_index_offset..).take(pieces.count()).collect(),
-                        pieces,
-                    };
-                    if !on_pieces_to_plot(pieces_to_plot) {
-                        // No need to continue
-                        break;
-                    }
+            let object_mapping = create_global_object_mapping(piece_index_offset, object_mapping);
 
-                    let object_mapping =
-                        create_global_object_mapping(piece_index_offset, object_mapping);
+            if let Err(error) = object_mappings.store(&object_mapping) {
+                error!(%error, "Failed to store object mappings for pieces");
+            }
 
-                    if let Err(error) = object_mappings.store(&object_mapping) {
-                        error!(%error, "Failed to store object mappings for pieces");
+            info!(segment_index, "Plotted segment");
+
+            ControlFlow::Continue(())
+        };
+
+        let archiving_handle = if let Some(node) = node {
+            tracing::trace!("Subscribing to pubsub archiving...");
+            let mut subscription = node
+                .subscribe(subspace_networking::PUB_SUB_ARCHIVING_TOPIC.clone())
+                .await?;
+            // TODO: Make sure this task can be cancelled
+            tokio::spawn(async move {
+                // TODO: add graceful handling for overload
+                while let Some(bytes) = subscription.next().await {
+                    match Decode::decode(&mut bytes.as_ref()) {
+                        Ok(archived_segment) => {
+                            if plot_segment(archived_segment) == ControlFlow::Break(()) {
+                                break;
+                            }
```

Comment:
Now this blocking call is called from withing async runtime. I think it might have been better in the end to leave blocking task the way it was before and just create two ways to call `archived_segments_sync_sender.send()`: one using RPC and another using DSN. I'll resolve cancellation TODO.

By changing `archived_segments_sync_sender`'s channel to bounded channel with some small capacity (every segment is just over 1M with all the metadata) you can replace `archived_segments_sync_sender.send()` with `archived_segments_sync_sender.try_send()` in DSN case and print warning when farmer isn't fast enough without running out of memory, fixing second TODO here.